watertap/property_models/ion_DSPMDE_prop_pack.py

```diff
@@ -0,0 +1,838 @@
+                                       block will deal with fixing/unfixing.
+            solver : Solver object to use during initialization if None is provided
+                     it will use the default solver for IDAES (default = None)
+            hold_state : flag indicating whether the initialization routine
+                         should unfix any state variables fixed during
+                         initialization (default=False).
+                         - True - states variables are not unfixed, and
+                                 a dict of returned containing flags for
+                                 which states were fixed during
+                                 initialization.
+                        - False - state variables are unfixed after
+                                 initialization by calling the
+                                 release_state method
+        Returns:
+            If hold_states is True, returns a dict containing flags for
+            which states were fixed during initialization.
+        """
+        # Get loggers
+        init_log = idaeslog.getInitLogger(self.name, outlvl, tag="properties")
+        solve_log = idaeslog.getSolveLogger(self.name, outlvl, tag="properties")
+
+        # Set solver and options
+        opt = get_solver(solver, optarg)
+
+        # Fix state variables
+        flags = fix_state_vars(self, state_args)
+        # Check when the state vars are fixed already result in dof 0
+        for k in self.keys():
+            dof = degrees_of_freedom(self[k])
+            if dof != 0:
+                raise PropertyPackageError("\nWhile initializing {sb_name}, the degrees of freedom "
+                                           "are {dof}, when zero is required. \nInitialization assumes "
+                                           "that the state variables should be fixed and that no other "
+                                           "variables are fixed. \nIf other properties have a "
+                                           "predetermined value, use the calculate_state method "
+                                           "before using initialize to determine the values for "
+                                           "the state variables and avoid fixing the property variables."
+                                           "".format(sb_name=self.name, dof=dof))
+
+        # ---------------------------------------------------------------------
+        skip_solve = True  # skip solve if only state variables are present
+        for k in self.keys():
+            if number_unfixed_variables(self[k]) != 0:
+                skip_solve = False
+
+        if not skip_solve:
+            # Initialize properties
+            with idaeslog.solver_log(solve_log, idaeslog.DEBUG) as slc:
+                results = solve_indexed_blocks(opt, [self], tee=slc.tee)
+            init_log.info_high("Property initialization: {}.".format(idaeslog.condition(results)))
+
```

Comment:
You should also add a check here that the final solve is optimal (or `None` if skipped), and raise an `InitializationError` if this is not the case.

Reply:
We added `fail_flag` to a couple utilities (`check_solve`) to provide the option of only providing a warning (when desired) as opposed to an exception. What do you think about adding `InitializationError` to the utility functions in util/initialization instead?

Reply:
The IDAES standard is now to raise an `InitializationError` here - the flowsheet can then choose to ignore it if it wants (with a `try/except`). The point was to make sure the end-user was aware of the issue and acknowledge it (by either fixing the problem or doing the `try/except`).

The problem with the utilities is it starts to bury this behaviour deeper so that it is hidden and less obvious what is happening.